strawberry/extensions/__init__.py

```diff
@@ -7,26 +7,28 @@ class Extension:
     def __init__(self, *, execution_context: ExecutionContext):
         self.execution_context = execution_context
 
-    def on_request_start(self):
-        ...
+    def on_request_start(self) -> Union[None, Awaitable[None]]:
+        """This method is called when a GraphQL request starts"""
 
-    def on_request_end(self):
-        ...
+    def on_request_end(self) -> Union[None, Awaitable[None]]:
+        """This method is called when a GraphQL request ends"""
 
-    def on_validation_start(self):
-        ...
+    def on_validation_start(self) -> Union[None, Awaitable[None]]:
+        """This method is called before the validation step"""
 
-    def on_validation_end(self):
-        ...
+    def on_validation_end(self) -> Union[None, Awaitable[None]]:
+        """This method is called after the validation step"""
 
-    def on_parsing_start(self):
-        ...
+    def on_parsing_start(self) -> Union[None, Awaitable[None]]:
+        """This method is called before the parsing step"""
 
-    def on_parsing_end(self):
-        ...
+    def on_parsing_end(self) -> Union[None, Awaitable[None]]:
+        """This method is called after the parsing step"""
 
-    def resolve(self, _next, root, info: Info, *args, **kwargs):
+    def resolve(
+        self, _next, root, info: Info, *args, **kwargs
+    ) -> Union[None, Awaitable[None]]:
         return _next(root, info, *args, **kwargs)
 
-    def get_results(self) -> Dict[str, Any]:
+    def get_results(self) -> Union[Dict[str, Any], Awaitable[Dict[str, Any]]]:
         return {}
```

Comment:
I wonder if we could do something like this to make the annotations in this file a bit more clear.
```python
T = TypeVar("T")
OptionallyAwaitable: TypeAlias = Union[T, Awaitable[T]]

def on_parsing_end(self) -> OptionallyAwaitable[None]:
    ...

def get_results(self) -> OptionallyAwaitable[Dict[str, object]]:
    ...
```

Reply:
We can do this in another PR 😊 

graphql-core has https://github.com/graphql-python/graphql-core/blob/03c2b06e4636ed8a89c7a4c7e56d244fd9d00bde/src/graphql/pyutils/awaitable_or_value.py

Reply:
I've done it :D